docs/design/mono/wasm-threads.md

```diff
@@ -0,0 +1,231 @@
+- Blazor renderBatch will continue working even with legacy interop in place.
+    - Because it only reads memory and it doesn't call back to Mono VM.
+- throw PNSE from Blazor's [`IJSInProcessRuntime.Invoke`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.jsinterop.ijsinprocessruntime.invoke)
+- throw PNSE from Blazor's any call to [`IJSUnmarshalledRuntime `](https://learn.microsoft.com/en-us/dotnet/api/microsoft.jsinterop.ijsunmarshalledruntime)
+
+# Current state 2023 Sep
+ - we already ship MT version of the runtime in the wasm-tools workload.
+ - It's enabled by `<WasmEnableThreads>true</WasmEnableThreads>` and it requires COOP HTTP headers.
+ - It will serve extra file `dotnet.native.worker.js`.
+ - This will also start in Blazor project, but UI rendering would not work.
+ - we have pre-allocated pool of browser Web Workers which are mapped to pthread dynamically.
+ - we can configure pthread to keep running after synchronous thread_main finished. That's necessary to run any async tasks involving JavaScript interop.
+ - GC is running on UI thread/worker.
+ - legacy interop has problems with GC boundaries.
+ - JSImport & JSExport work
+ - There is private JSSynchronizationContext implementation which is too synchronous
+ - There is draft of public C# API for creating JSWebWorker with JS interop. It must be dedicated un-managed resource, because we could not cleanup JS state created by user code.
+ - There is MT version of HTTP & WS clients, which could be called from any thread but it's also too synchronous implementation.
+ - Many unit tests fail on MT https://github.com/dotnet/runtime/pull/91536
+ - there are MT C# ref assemblies, which don't throw PNSE for MT build of the runtime for blocking APIs.
+
+## Task breakdown
+- [ ] rename `WebWorker` API to `JSWebWorker` ?
+- [ ] design details of JSImport binding, allocation, asynchrony
+- [ ] design details of JSExport binding, allocation, asynchrony
+- [ ] `ToManaged(out Task)` to be called before the actual JS method
+- [ ] public API for `JSHost.<Target>SynchronizationContext` which could be used by code generator.
+- [ ] change the code gen for JSImport
+- [ ] change the code gen for JSExport
+- [ ] reimplement `JSSynchronizationContext` to be more async
+- [ ] implement Blazor's `WebAssemblyDispatcher`
+- [ ] reimplement HTTP and WS with the new code gen without direct SynchronizationContext use
+    - [ ] there is synchronous callback from JS event to C# in HTTP code.
+- [ ] make C# finalizers work
+- [ ] throw PNSE - fail fast, so that users discover limits in the dev loop
+    - [ ] on any MT use of `mono_bind_static_method` from legacy interop.
+        - Because it's synchronous. It throws on JSWebWorker already.
+    - [ ] on UI synchronous JSImport
+    - [ ] on UI synchronous C# delegate callback
+    - [ ] throw fatal if somehow C# code was blocking on UI thread.
+- [ ] optinal: make underlying emscripten WebWorker pool allocation dynamic
```

Comment:
we could also expose an async managed API to spin up additional workers (there's a GH Issue for that somewhere...)